tests/net/dhcpv4/src/main.c

```diff
@@ -203,128 +203,40 @@ static void net_dhcpv4_iface_init(struct net_if *iface)
 	net_if_set_link_addr(iface, mac, 6, NET_LINK_ETHERNET);
 }
 
-static struct net_buf *pkt_get_data(struct net_pkt *pkt, struct net_if *iface)
-{
-	struct net_buf *frag;
-	struct net_eth_hdr *hdr;
-
-	frag = net_pkt_get_frag(pkt, K_FOREVER);
-	if (!frag) {
-		return NULL;
-	}
-
-	net_buf_add(frag, sizeof(struct net_eth_hdr));
-	net_pkt_frag_add(pkt, frag);
-
-	hdr = (struct net_eth_hdr *)(net_pkt_data(pkt));
-	hdr->type = htons(NET_ETH_PTYPE_IP);
-
-	net_ipaddr_copy(&hdr->dst, &src_addr);
-	net_ipaddr_copy(&hdr->src, &dst_addr);
-
-	return frag;
-}
-
-static void set_ipv4_header(struct net_pkt *pkt)
-{
-	struct net_ipv4_hdr *ipv4;
-	u16_t length;
-
-	ipv4 = NET_IPV4_HDR(pkt);
-
-	ipv4->vhl = 0x45; /* IP version and header length */
-	ipv4->tos = 0x00;
-
-	length = sizeof(offer) + sizeof(struct net_ipv4_hdr) +
-		 sizeof(struct net_udp_hdr);
-
-	ipv4->len = htons(length);
-
-	(void)memset(ipv4->id, 0, 4); /* id and offset */
-
-	ipv4->ttl = 0xFF;
-	ipv4->proto = IPPROTO_UDP;
-
-	net_ipaddr_copy(&ipv4->src, &server_addr);
-	net_ipaddr_copy(&ipv4->dst, &client_addr);
-}
-
-static void set_udp_header(struct net_pkt *pkt)
-{
-	struct net_udp_hdr *udp;
-	u16_t length;
-
-	udp = (struct net_udp_hdr *)((u8_t *)(NET_IPV4_HDR(pkt)) +
-				     sizeof(struct net_ipv4_hdr));
-
-	udp->src_port = htons(SERVER_PORT);
-	udp->dst_port = htons(CLIENT_PORT);
-
-	length = sizeof(offer) + sizeof(struct net_udp_hdr);
-	udp->len = htons(length);
-	udp->chksum = 0;
-}
-
 struct net_pkt *prepare_dhcp_offer(struct net_if *iface, u32_t xid)
 {
 	struct net_pkt *pkt;
-	struct net_buf *frag;
-	int bytes, remaining = sizeof(offer), pos = 0;
-	u16_t offset;
-
-	pkt = net_pkt_get_reserve_rx(K_FOREVER);
-	if (!pkt) {
-		return NULL;
-	}
-
-	frag = pkt_get_data(pkt, iface);
-	if (!frag) {
-		net_pkt_unref(pkt);
-		return NULL;
-	}
-
-	net_pkt_set_iface(pkt, iface);
-	net_pkt_set_family(pkt, AF_INET);
-	net_pkt_set_ip_hdr_len(pkt, sizeof(struct net_ipv4_hdr));
-
-	/* Place the IPv4 header */
-	set_ipv4_header(pkt);
-
-	/* Place the UDP header */
-	set_udp_header(pkt);
 
-	net_buf_add(frag, NET_IPV4UDPH_LEN);
-	offset = NET_IPV4UDPH_LEN;
+	pkt = net_pkt_alloc_with_buffer(iface,
+					sizeof(struct net_ipv4_hdr) +
```

Comment:
When given family is not AF_UNSPEC, the allocator will take care of counting the header size, same for protocol if it's not 0. So you can remove the ipv4 and udp hdr size

Apply that where relevant

Reply:
PR merged already, I will do that in separate PR.

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/13946 fixes issues you mentioned.